src/Data/Text/Utf8.hs

```diff
@@ -245,34 +205,12 @@ decodeUtf8 (cu0 : cu1 : cu2 : cus) | cu0 < 0xf0 = decode3 cu0 cu1 cu2 : decodeUt
 decodeUtf8 (cu0 : cu1 : cu2 : cu3 : cus) | cu0 < 0xf8 = decode4 cu0 cu1 cu2 cu3 : decodeUtf8 cus
 decodeUtf8 cus = error $ "Invalid UTF-8 input sequence at " ++ show (take 4 cus)
```

Comment:
It's likely faster to use [`utf8LengthByLeader`](https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Internal-Encoding-Utf8.html#v:utf8LengthByLeader) and then switch between `chr2`, `chr3`, `chr4`.